File: Problem14.py
def valuefinder(d, key):
	if(d.has_key(key)):
		return d[key]
	else:
	 	return -1

collatz = {1: 1}
num = 1
maxcount = 1

for i in range(2,3):
	seq = i
	count = 1
	while(1):
		if(seq==1):
			break
		elif(valuefinder(collatz, seq)==-1):
			if(seq%2==0):
				seq /= 2
				count+=1
				if(seq==1):
					collatz.update({i: count})
					if(count>maxcount):
						maxcount = count
						num = i
			elif(seq%2==1):
				seq = 3*seq+1
				count+=1
		else:
			count+= valuefinder(collatz, seq)
			if(seq==1):
				if(count>maxcount):
					num = i
				break

print num, maxcount

File: Problem14.c
#include<stdio.h>

int collatz(long int num) {
	int count = 0;
	while(num!=1) {
		if(num%2==0) {
			num/=2;
			count++;
		}
		else {
			num = 3*num + 1;
			count++;
		}
	}
	return count;
}

int main() {
	long int i, num;
	int max = 1;
	for(i=1; i<46500; i++) {
		int count = collatz(i);
		if(count>max) {
			num = i;
			max = count;
		}
	}
	printf("%li -> %d\n", num, max);
	return 0;
}
